rpc: add missing fields to wallet rpc help output

ariard opened the pull request:
Knowledge of `walletconflicts` field existence is really nice when you're debugging conflicts. Was added in #3671 but never documented in RPC helps.

Others were added after a quick skim, we may still have missing ones in wallet rpcs.

fanquake commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1396,11 +1397,13 @@ UniValue listtransactions(const JSONRPCRequest& request)
             "    \"blockindex\": n,          (numeric) The index of the transaction in the block that includes it.\n"
             "    \"blocktime\": xxx,         (numeric) The block time in seconds since epoch (1 Jan 1970 GMT).\n"
             "    \"txid\": \"transactionid\", (string) The transaction id.\n"
+            "    \"walletconflicts:\"hashvalue\", (array) List of txid with which this transaction is conflicting.\n"
```
nit: missing a space after the `:` here.

fanquake commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1381,6 +1381,7 @@ UniValue listtransactions(const JSONRPCRequest& request)
                 RPCResult{
             "[\n"
             "  {\n"
+            "    \"involvesWatchonly\": xxx, (bool) Only returned if imported addresses were involved in transaction\n"
```
Matches the description from the `listreceivedbylabel` [help](https://github.com/bitcoin/bitcoin/blob/master/src/wallet/rpcwallet.cpp#L1243).

MarcoFalke replied:
I'd rather pull this out into a static const RPCResult in a moveonly commit, since it is used in three places. And then only update that single place with the documentation. We do that in other places (like getmempoolentry)

promag replied:
Agree.

fanquake reviewed (commented):
Concept ACK - I'll have a look if there are more missing fields.

`walletconflicts` was added in #3671.

nit: In the commit body I assume you mean `gettransaction`.

emilengler commented:
Concept ACK

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1396,11 +1397,13 @@ UniValue listtransactions(const JSONRPCRequest& request)
             "                                         'send' category of transactions.\n"
             "    \"confirmations\": n,       (numeric) The number of confirmations for the transaction. Negative confirmations indicate the\n"
             "                                         transaction conflicts with the block chain\n"
+            "    \"generated\": xxx,         (bool) Only returns true if transaction is a coinbase one\n"
```
The coinbase is the input. The output is simply generated.

ariard replied:
Doesn't a transaction including a coinbase input qualify as a coinbase one ? We don't enforce this wording strictly if the rest of the codebase I think.

Commit:
MOVEONLY : move RPC wallets helpers to TransactionDescriptionString

Commit:
Add missing fields in TransactionDescriptionString and others

Fields involvesWatchonly, generated, walletconflicts were missing
in result description of listtransactions, listsinceblock,
gettransaction

Align getttransaction fields which were odd compare to other rpc
helpers

ariard commented:
Updated at fbc3c9f as suggested in https://github.com/bitcoin/bitcoin/pull/16689#discussion_r317160287 (but only output from WalletTxToJSON as we may have small diffs between RPCs for other fields)

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1393,19 +1411,8 @@ UniValue listtransactions(const JSONRPCRequest& request)
             "    \"label\": \"label\",       (string) A comment for the address/transaction, if any\n"
             "    \"vout\": n,                (numeric) the vout value\n"
             "    \"fee\": x.xxx,             (numeric) The amount of the fee in " + CURRENCY_UNIT + ". This is negative and only available for the \n"
-            "                                         'send' category of transactions.\n"
```
This line should not be deleted.

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1359,6 +1359,23 @@ static void ListTransactions(interfaces::Chain::Lock& locked_chain, CWallet* con
+static const std::string TransactionDescriptionString()
+{
+    return "    \"confirmations\": n,       (numeric) The number of confirmations for the transaction. Negative confirmations indicate the\n"
+           "                                         transaction conflicts with the block chain\n"
```
While extracting this, it wouldn't hurt to fixup the nits that wouldn't be worth a commit on their own, like adding the missing period at the end of this sentence and next sentence.

jonatack replied:
e.g. s/chain\n"/chain.\n"/ and s/one\n"/one.\n"/

jonatack replied:
Suggestion: s/Only returns true if transaction is a coinbase one./True if the transaction is a coinbase, otherwise false./

jonatack replied:
It would be nice to align the horizontal spacing of this entire section.

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1359,6 +1359,23 @@ static void ListTransactions(interfaces::Chain::Lock& locked_chain, CWallet* con
+           "    \"blockindex\": n,          (numeric) The index of the transaction in the block that includes it.\n"
+           "    \"blocktime\": xxx,         (numeric) The block time in seconds since epoch (1 Jan 1970 GMT).\n"
+           "    \"txid\": \"transactionid\", (string) The transaction id.\n"
+           "    \"walletconflicts: \"hashvalue\", (array) List of txid with which this transaction is conflicting.\n"
```
Grammar fix: s/List of txid with which this transaction is conflicting./List of txids that conflict with this transaction./

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1393,19 +1411,8 @@ UniValue listtransactions(const JSONRPCRequest& request)
-            "    \"bip125-replaceable\": \"yes|no|unknown\",  (string) Whether this transaction could be replaced due to BIP125 (replace-by-fee);\n"
-            "                                                     may be unknown for unconfirmed transactions not in the mempool\n"
+	    + TransactionDescriptionString()
+            + "                                         'send' category of transactions.\n"
```
This line should be removed/moved to before TransactionDescriptionString) -- it's in the wrong place.

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1515,6 +1522,7 @@ static UniValue listsinceblock(const JSONRPCRequest& request)
                 RPCResult{
             "{\n"
             "  \"transactions\": [\n"
+            "    \"involvesWatchonly\": xxx, (bool) Only returned if imported addresses were involved in transaction\n"
```
"Only returned" means what? Maybe it's just me, but this and the rest of the sentence isn't clear.

nit: missing period at end of sentence.

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1655,17 +1654,10 @@ static UniValue gettransaction(const JSONRPCRequest& request)
+	    + TransactionDescriptionString()
+            + "  \"details\" : [\n"
             "    {\n"
+            "      \"involvesWatchonly\": xxx,         (bool) Only returned if imported addresses were involved in transaction\n"
```
Same comment as for the identical line above.

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1526,17 +1534,8 @@ static UniValue listsinceblock(const JSONRPCRequest& request)
-            "    \"bip125-replaceable\": \"yes|no|unknown\",  (string) Whether this transaction could be replaced due to BIP125 (replace-by-fee);\n"
-            "                                                   may be unknown for unconfirmed transactions not in the mempool\n"
-            "    \"abandoned\": xxx,         (bool) 'true' if the transaction has been abandoned (inputs are respendable). Only available for the 'send' category of transactions.\n"
+	    + TransactionDescriptionString()
```
nit: It looks like you added a tab at the start of this line.

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1393,19 +1411,8 @@ UniValue listtransactions(const JSONRPCRequest& request)
-            "    \"comment\": \"...\",       (string) If a comment is associated with the transaction.\n"
-            "    \"bip125-replaceable\": \"yes|no|unknown\",  (string) Whether this transaction could be replaced due to BIP125 (replace-by-fee);\n"
-            "                                                     may be unknown for unconfirmed transactions not in the mempool\n"
+	    + TransactionDescriptionString()
```
nit: It looks like you've added a tab at the start of this line.

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1655,17 +1654,10 @@ static UniValue gettransaction(const JSONRPCRequest& request)
-            "  \"bip125-replaceable\": \"yes|no|unknown\",  (string) Whether this transaction could be replaced due to BIP125 (replace-by-fee);\n"
-            "                                                   may be unknown for unconfirmed transactions not in the mempool\n"
-            "  \"details\" : [\n"
+	    + TransactionDescriptionString()
```
nit: another tab.

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1526,17 +1534,8 @@ static UniValue listsinceblock(const JSONRPCRequest& request)
-            "                                                   may be unknown for unconfirmed transactions not in the mempool\n"
-            "    \"abandoned\": xxx,         (bool) 'true' if the transaction has been abandoned (inputs are respendable). Only available for the 'send' category of transactions.\n"
+	    + TransactionDescriptionString()
+            + "    \"abandoned\": xxx,         (bool) 'true' if the transaction has been abandoned (inputs are respendable). Only available for the 'send' category of transactions.\n"
```
s/'true'/True/

Possibly: s/Only available for/Only applies to/

jonatack reviewed (commented):
Concept ACK.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1359,6 +1359,23 @@ static void ListTransactions(interfaces::Chain::Lock& locked_chain, CWallet* con
+           "                                                       transaction conflicts with the block chain.\n"
+           "    \"generated\": xxx,                          (bool) Only returns true if transaction only input is a coinbase one, otherwise false.\n"
+           "    \"trusted\": xxx,                            (bool) Whether we consider the outputs of this unconfirmed transaction safe to spend.\n"
+           "    \"blockhash\": \"hashvalue\",                (string) The block hash containing the transaction.\n"
```
nit, try to have it aligned with surrounding lines.

ariard replied:
You mean `blockhash` compare to `trusted` ?

MarcoFalke replied:
I think the first commit (moveonly) should actually be moveonly (via `git diff --color-moved=dimmed-zebra`).

Otherwise the rendered whitespace will be messed up. (You can get the help text with the `help listtransactions` rpc)

meshcollider commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1394,19 +1409,8 @@ UniValue listtransactions(const JSONRPCRequest& request)
-            "    \"txid\": \"transactionid\", (string) The transaction id.\n"
-            "    \"time\": xxx,              (numeric) The transaction time in seconds since epoch (midnight Jan 1 1970 GMT).\n"
-            "    \"timereceived\": xxx,      (numeric) The time received in seconds since epoch (midnight Jan 1 1970 GMT).\n"
-            "    \"comment\": \"...\",       (string) If a comment is associated with the transaction.\n"
```
Not sure if this was a deliberate removal but `comment` is still a field which might be present, among others.
```
for (const std::pair<const std::string, std::string>& item : wtx.mapValue)
        entry.pushKV(item.first, item.second);
```
(From WalletTxToJSON)

promag replied:
Good catch, could mention that it is only present if not empty.

fanquake replied:
@ariard Could you follow up here?

ariard replied:
Sorry for delay, pushed correction

kristapsk commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1359,6 +1359,24 @@ static void ListTransactions(interfaces::Chain::Lock& locked_chain, CWallet* con
+           "    \"blockindex\": n,                           (numeric) The index of the transaction in the block that includes it.\n"
+           "    \"blocktime\": xxx,                          (numeric) The block time in seconds since epoch (1 Jan 1970 GMT).\n"
+           "    \"txid\": \"transactionid\",                 (string) The transaction id.\n"
+           "    \"walletconflicts: \"hashvalue\",            (array) List of txids that conflict with this transaction.\n"
```
It returns array here. So `\"walletconflicts\": [ \"TXID\", ... ]` should be used instead. See #16869 as an example.

MarcoFalke commented:
<details><summary>Rendered diff:</summary>

```diff
diff --git a/gettransaction b/gettransaction
index cf09efd..7ef6106 100644
--- a/gettransaction
+++ b/gettransaction
@@ -10,37 +10,46 @@ Arguments:
 Result:
 {
   "amount" : x.xxx,        (numeric) The transaction amount in BTC
   "fee": x.xxx,            (numeric) The amount of the fee in BTC. This is negative and only available for the 
                               'send' category of transactions.
-  "confirmations" : n,     (numeric) The number of confirmations
-  "blockhash" : "hash",  (string) The block hash
-  "blockindex" : xx,       (numeric) The index of the transaction in the block that includes it
-  "blocktime" : ttt,       (numeric) The time in seconds since epoch (1 Jan 1970 GMT)
+    "confirmations": n,                        (numeric) The number of confirmations for the transaction. Negative confirmations indicate the
+                                                       transaction conflicts with the block chain.
+    "generated": xxx,                          (bool) Only returns true if transaction only input is a coinbase one, otherwise false.
+    "trusted": xxx,                            (bool) Whether we consider the outputs of this unconfirmed transaction safe to spend.
+    "blockhash": "hashvalue",                  (string) The block hash containing the transaction.
+    "blockindex": n,                           (numeric) The index of the transaction in the block that includes it.
+    "blocktime": xxx,                          (numeric) The block time in seconds since epoch (1 Jan 1970 GMT).
     "txid": "transactionid",                   (string) The transaction id.
-  "time" : ttt,            (numeric) The transaction time in seconds since epoch (1 Jan 1970 GMT)
-  "timereceived" : ttt,    (numeric) The time received in seconds since epoch (1 Jan 1970 GMT)
+    "walletconflicts": [                       (array) List of txids.
+      "txid",                                  (string) The conflicting txid.
+      ...
+    ],
+    "time": xxx,                               (numeric) The transaction time in seconds since epoch (midnight Jan 1 1970 GMT).
+    "timereceived": xxx,                       (numeric) The time received in seconds since epoch (midnight Jan 1 1970 GMT).
+    "comment": "...",                          (string) If a comment is associated with the transaction, only present if not empty.
     "bip125-replaceable": "yes|no|unknown",    (string) Whether this transaction could be replaced due to BIP125 (replace-by-fee);
                                                      may be unknown for unconfirmed transactions not in the mempool
   "details" : [
     {
+      "involvesWatchonly": xxx,         (bool) Only returns true if imported addresses were involved in transaction.
       "address" : "address",          (string) The bitcoin address involved in the transaction
       "category" :                      (string) The transaction category.
                    "send"                  Transactions sent.
                    "receive"               Non-coinbase transactions received.
                    "generate"              Coinbase transactions received with more than 100 confirmations.
                    "immature"              Coinbase transactions received with 100 or fewer confirmations.
                    "orphan"                Orphaned coinbase transactions received.
       "amount" : x.xxx,                 (numeric) The amount in BTC
       "label" : "label",              (string) A comment for the address/transaction, if any
       "vout" : n,                       (numeric) the vout value
       "fee": x.xxx,                     (numeric) The amount of the fee in BTC. This is negative and only available for the 
                                            'send' category of transactions.
       "abandoned": xxx                  (bool) 'true' if the transaction has been abandoned (inputs are respendable). Only available for the 
                                            'send' category of transactions.
     }
     ,...
   ],
   "hex" : "data"         (string) Raw data for transaction
   "decoded" : transaction         (json object) Optional, the decoded transaction (only present when `verbose` is passed), equivalent to the
                                                 RPC decoderawtransaction method, or the RPC getrawtransaction method when `verbose` is passed.
 }
diff --git a/listreceivedbyaddress b/listreceivedbyaddress
index 3e895f4..566ddb3 100644
--- a/listreceivedbyaddress
+++ b/listreceivedbyaddress
@@ -11,15 +11,15 @@ Arguments:
 Result:
 [
   {
-    "involvesWatchonly" : true,        (bool) Only returned if imported addresses were involved in transaction
+    "involvesWatchonly" : true,        (bool) Only returns true if imported addresses were involved in transaction.
     "address" : "receivingaddress",  (string) The receiving address
     "amount" : x.xxx,                  (numeric) The total amount in BTC received by the address
     "confirmations" : n,               (numeric) The number of confirmations of the most recent transaction included
     "label" : "label",               (string) The label of the receiving address. The default label is "".
     "txids": [
        "txid",                         (string) The ids of transactions received with the address 
        ...
     ]
   }
   ,...
 ]
diff --git a/listreceivedbylabel b/listreceivedbylabel
index 38d25bf..d716f9b 100644
--- a/listreceivedbylabel
+++ b/listreceivedbylabel
@@ -10,10 +10,10 @@ Arguments:
 Result:
 [
   {
-    "involvesWatchonly" : true,   (bool) Only returned if imported addresses were involved in transaction
+    "involvesWatchonly" : true,   (bool) Only returns true if imported addresses were involved in transaction.
     "amount" : x.xxx,             (numeric) The total amount received by addresses with this label
     "confirmations" : n,          (numeric) The number of confirmations of the most recent transaction included
     "label" : "label"           (string) The label of the receiving address. The default label is "".
   }
   ,...
 ]
diff --git a/listsinceblock b/listsinceblock
index be347aa..9a9b8f7 100644
--- a/listsinceblock
+++ b/listsinceblock
@@ -14,35 +14,43 @@ Arguments:
 Result:
 {
   "transactions": [
+    "involvesWatchonly": xxx, (bool) Only returns true if imported addresses were involved in transaction.
     "address":"address",    (string) The bitcoin address of the transaction.
     "category":               (string) The transaction category.
                 "send"                  Transactions sent.
                 "receive"               Non-coinbase transactions received.
                 "generate"              Coinbase transactions received with more than 100 confirmations.
                 "immature"              Coinbase transactions received with 100 or fewer confirmations.
                 "orphan"                Orphaned coinbase transactions received.
     "amount": x.xxx,          (numeric) The amount in BTC. This is negative for the 'send' category, and is positive
                                          for all other categories
     "vout" : n,               (numeric) the vout value
     "fee": x.xxx,             (numeric) The amount of the fee in BTC. This is negative and only available for the 'send' category of transactions.
-    "confirmations": n,       (numeric) The number of confirmations for the transaction.
-                                          When it's < 0, it means the transaction conflicted that many blocks ago.
+    "confirmations": n,                        (numeric) The number of confirmations for the transaction. Negative confirmations indicate the
+                                                       transaction conflicts with the block chain.
+    "generated": xxx,                          (bool) Only returns true if transaction only input is a coinbase one, otherwise false.
+    "trusted": xxx,                            (bool) Whether we consider the outputs of this unconfirmed transaction safe to spend.
     "blockhash": "hashvalue",                  (string) The block hash containing the transaction.
     "blockindex": n,                           (numeric) The index of the transaction in the block that includes it.
     "blocktime": xxx,                          (numeric) The block time in seconds since epoch (1 Jan 1970 GMT).
     "txid": "transactionid",                   (string) The transaction id.
-    "time": xxx,              (numeric) The transaction time in seconds since epoch (Jan 1 1970 GMT).
-    "timereceived": xxx,      (numeric) The time received in seconds since epoch (Jan 1 1970 GMT).
+    "walletconflicts": [                       (array) List of txids.
+      "txid",                                  (string) The conflicting txid.
+      ...
+    ],
+    "time": xxx,                               (numeric) The transaction time in seconds since epoch (midnight Jan 1 1970 GMT).
+    "timereceived": xxx,                       (numeric) The time received in seconds since epoch (midnight Jan 1 1970 GMT).
+    "comment": "...",                          (string) If a comment is associated with the transaction, only present if not empty.
     "bip125-replaceable": "yes|no|unknown",    (string) Whether this transaction could be replaced due to BIP125 (replace-by-fee);
                                                      may be unknown for unconfirmed transactions not in the mempool
     "abandoned": xxx,         (bool) 'true' if the transaction has been abandoned (inputs are respendable). Only available for the 'send' category of transactions.
     "comment": "...",       (string) If a comment is associated with the transaction.
     "label" : "label"       (string) A comment for the address/transaction, if any
     "to": "...",            (string) If a comment to is associated with the transaction.
   ],
   "removed": [
     <structure is the same as "transactions" above, only present if include_removed=true>
     Note: transactions that were re-added in the active chain will appear as-is in this array, and may thus have a positive confirmation count.
   ],
   "lastblock": "lastblockhash"     (string) The hash of the block (target_confirmations-1) from the best block on the main chain. This is typically used to feed back into listsinceblock the next time you call it. So you would generally use a target_confirmations of say 6, so you will be continually re-notified of transactions until they've reached 6 confirmations plus any new ones
 }
diff --git a/listtransactions b/listtransactions
index 438677e..ef584d4 100644
--- a/listtransactions
+++ b/listtransactions
@@ -14,32 +14,38 @@ Arguments:
 Result:
 [
   {
+    "involvesWatchonly": xxx, (bool) Only returns true if imported addresses were involved in transaction.
     "address":"address",    (string) The bitcoin address of the transaction.
     "category":               (string) The transaction category.
                 "send"                  Transactions sent.
                 "receive"               Non-coinbase transactions received.
                 "generate"              Coinbase transactions received with more than 100 confirmations.
                 "immature"              Coinbase transactions received with 100 or fewer confirmations.
                 "orphan"                Orphaned coinbase transactions received.
     "amount": x.xxx,          (numeric) The amount in BTC. This is negative for the 'send' category, and is positive
                                         for all other categories
     "label": "label",       (string) A comment for the address/transaction, if any
     "vout": n,                (numeric) the vout value
     "fee": x.xxx,             (numeric) The amount of the fee in BTC. This is negative and only available for the 
                                          'send' category of transactions.
     "confirmations": n,                        (numeric) The number of confirmations for the transaction. Negative confirmations indicate the
-                                         transaction conflicts with the block chain
+                                                       transaction conflicts with the block chain.
+    "generated": xxx,                          (bool) Only returns true if transaction only input is a coinbase one, otherwise false.
     "trusted": xxx,                            (bool) Whether we consider the outputs of this unconfirmed transaction safe to spend.
     "blockhash": "hashvalue",                  (string) The block hash containing the transaction.
     "blockindex": n,                           (numeric) The index of the transaction in the block that includes it.
     "blocktime": xxx,                          (numeric) The block time in seconds since epoch (1 Jan 1970 GMT).
     "txid": "transactionid",                   (string) The transaction id.
+    "walletconflicts": [                       (array) List of txids.
+      "txid",                                  (string) The conflicting txid.
+      ...
+    ],
     "time": xxx,                               (numeric) The transaction time in seconds since epoch (midnight Jan 1 1970 GMT).
     "timereceived": xxx,                       (numeric) The time received in seconds since epoch (midnight Jan 1 1970 GMT).
-    "comment": "...",       (string) If a comment is associated with the transaction.
+    "comment": "...",                          (string) If a comment is associated with the transaction, only present if not empty.
     "bip125-replaceable": "yes|no|unknown",    (string) Whether this transaction could be replaced due to BIP125 (replace-by-fee);
                                                      may be unknown for unconfirmed transactions not in the mempool
     "abandoned": xxx          (bool) 'true' if the transaction has been abandoned (inputs are respendable). Only available for the 
                                          'send' category of transactions.
   }
 ]

```

</details>

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1359,6 +1359,27 @@ static void ListTransactions(interfaces::Chain::Lock& locked_chain, CWallet* con
+{
+    return "    \"confirmations\": n,                        (numeric) The number of confirmations for the transaction. Negative confirmations indicate the\n"
+           "                                                       transaction conflicts with the block chain.\n"
+           "    \"generated\": xxx,                          (bool) Only returns true if transaction only input is a coinbase one, otherwise false.\n"
```
This is wrong. The field can never be `false`

kristapsk commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1363,11 +1363,16 @@ static const std::string TransactionDescriptionString()
            "    \"blocktime\": xxx,                          (numeric) The block time in seconds since epoch (1 Jan 1970 GMT).\n"
            "    \"txid\": \"transactionid\",                   (string) The transaction id.\n"
+           "    \"walletconflicts\": [                       (array) List of txids.\n"
+           "      \"txid\",                                  (string) The conflicting txid.\n"
```
I think this would be better describing (based on #16869):
```
            "    \"walletconflicts\":[      (array) Conflicting transaction ids (double spends).\n"
            "        \"TXID\",              (string) The transaction id.\n"
```

fanquake commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1359,6 +1359,27 @@ static void ListTransactions(interfaces::Chain::Lock& locked_chain, CWallet* con
+    return "    \"confirmations\": n,                        (numeric) The number of confirmations for the transaction. Negative confirmations indicate the\n"
+           "                                                       transaction conflicts with the block chain.\n"
+           "    \"generated\": xxx,                          (bool) Only returns true if transaction only input is a coinbase one, otherwise false.\n"
+           "    \"trusted\": xxx,                            (bool) Whether we consider the outputs of this unconfirmed transaction safe to spend.\n"
```
Could you update this to be `(bool, optional)` and address [this comment](https://github.com/bitcoin/bitcoin/pull/16951#discussion_r327494296):

> if this is optional, I think the help text should also mention when the value is present or not.

ariard replied:
Rewrote comment but not sure about optional, it seems to be reserved when a optional feature is involved like HD-wallet, which is not the case with IsTrusted

laanwj commented:
@ariard If you'd like this to be included in 0.19, please make the suggested changes today or tomorrow, otherwise it will miss the split-off.

ariard commented:
Sorry for the delay, was offline for few days, updated with suggested changes at 396da1d

MarcoFalke commented:
Rendered diff:

```diff
diff --git a/gettransaction b/gettransaction
index cf09efd..84731c7 100644
--- a/gettransaction
+++ b/gettransaction
@@ -10,39 +10,47 @@ Arguments:
 Result:
 {
     "amount" : x.xxx,        (numeric) The transaction amount in BTC
     "fee": x.xxx,            (numeric) The amount of the fee in BTC. This is negative and only available for the 
                               'send' category of transactions.
-  "confirmations" : n,     (numeric) The number of confirmations
-  "blockhash" : "hash",  (string) The block hash
-  "blockindex" : xx,       (numeric) The index of the transaction in the block that includes it
-  "blocktime" : ttt,       (numeric) The time in seconds since epoch (1 Jan 1970 GMT)
+    "confirmations": n,                        (numeric) The number of confirmations for the transaction. Negative confirmations indicate the
+                                                       transaction conflicts with the block chain.
+    "generated": xxx,                          (bool) Only present if transaction only input is a coinbase one.
+    "trusted": xxx,                            (bool) Only present if we consider transaction to be trusted and so safe to spend from.
+    "blockhash": "hashvalue",                  (string) The block hash containing the transaction.
+    "blockindex": n,                           (numeric) The index of the transaction in the block that includes it.
+    "blocktime": xxx,                          (numeric) The block time in seconds since epoch (1 Jan 1970 GMT).
     "txid": "transactionid",                   (string) The transaction id.
-  "time" : ttt,            (numeric) The transaction time in seconds since epoch (1 Jan 1970 GMT)
-  "timereceived" : ttt,    (numeric) The time received in seconds since epoch (1 Jan 1970 GMT)
+    "walletconflicts": [                       (array) Conflicting transaction ids.
+      "txid",                                  (string) The transaction id.
+      ...
+    ],
+    "time": xxx,                               (numeric) The transaction time in seconds since epoch (midnight Jan 1 1970 GMT).
+    "timereceived": xxx,                       (numeric) The time received in seconds since epoch (midnight Jan 1 1970 GMT).
+    "comment": "...",                          (string) If a comment is associated with the transaction, only present if not empty.
     "bip125-replaceable": "yes|no|unknown",    (string) Whether this transaction could be replaced due to BIP125 (replace-by-fee);
                                                      may be unknown for unconfirmed transactions not in the mempool
     "details" : [
       {
+        "involvesWatchonly": xxx,         (bool) Only returns true if imported addresses were involved in transaction.
         "address" : "address",          (string) The bitcoin address involved in the transaction
         "category" :                      (string) The transaction category.
                      "send"                  Transactions sent.
                      "receive"               Non-coinbase transactions received.
                      "generate"              Coinbase transactions received with more than 100 confirmations.
                      "immature"              Coinbase transactions received with 100 or fewer confirmations.
                      "orphan"                Orphaned coinbase transactions received.
         "amount" : x.xxx,                 (numeric) The amount in BTC
         "label" : "label",              (string) A comment for the address/transaction, if any
         "vout" : n,                       (numeric) the vout value
         "fee": x.xxx,                     (numeric) The amount of the fee in BTC. This is negative and only available for the 
                                            'send' category of transactions.
         "abandoned": xxx                  (bool) 'true' if the transaction has been abandoned (inputs are respendable). Only available for the 
                                            'send' category of transactions.
       }
       ,...
     ],
     "hex" : "data"         (string) Raw data for transaction
     "decoded" : transaction         (json object) Optional, the decoded transaction (only present when `verbose` is passed), equivalent to the
-                                                RPC decoderawtransaction method, or the RPC getrawtransaction method when `verbose` is passed.
 }
 
 Examples:
diff --git a/listreceivedbyaddress b/listreceivedbyaddress
index 3e895f4..566ddb3 100644
--- a/listreceivedbyaddress
+++ b/listreceivedbyaddress
@@ -11,15 +11,15 @@ Arguments:
 Result:
 [
   {
-    "involvesWatchonly" : true,        (bool) Only returned if imported addresses were involved in transaction
+    "involvesWatchonly" : true,        (bool) Only returns true if imported addresses were involved in transaction.
     "address" : "receivingaddress",  (string) The receiving address
     "amount" : x.xxx,                  (numeric) The total amount in BTC received by the address
     "confirmations" : n,               (numeric) The number of confirmations of the most recent transaction included
     "label" : "label",               (string) The label of the receiving address. The default label is "".
     "txids": [
        "txid",                         (string) The ids of transactions received with the address 
        ...
     ]
   }
   ,...
 ]
diff --git a/listreceivedbylabel b/listreceivedbylabel
index 38d25bf..d716f9b 100644
--- a/listreceivedbylabel
+++ b/listreceivedbylabel
@@ -10,10 +10,10 @@ Arguments:
 Result:
 [
   {
-    "involvesWatchonly" : true,   (bool) Only returned if imported addresses were involved in transaction
+    "involvesWatchonly" : true,   (bool) Only returns true if imported addresses were involved in transaction.
     "amount" : x.xxx,             (numeric) The total amount received by addresses with this label
     "confirmations" : n,          (numeric) The number of confirmations of the most recent transaction included
     "label" : "label"           (string) The label of the receiving address. The default label is "".
   }
   ,...
 ]
diff --git a/listsinceblock b/listsinceblock
index be347aa..54f4fbf 100644
--- a/listsinceblock
+++ b/listsinceblock
@@ -14,35 +14,43 @@ Arguments:
 Result:
 {
   "transactions": [
+    "involvesWatchonly": xxx, (bool) Only returns true if imported addresses were involved in transaction.
     "address":"address",    (string) The bitcoin address of the transaction.
     "category":               (string) The transaction category.
                 "send"                  Transactions sent.
                 "receive"               Non-coinbase transactions received.
                 "generate"              Coinbase transactions received with more than 100 confirmations.
                 "immature"              Coinbase transactions received with 100 or fewer confirmations.
                 "orphan"                Orphaned coinbase transactions received.
     "amount": x.xxx,          (numeric) The amount in BTC. This is negative for the 'send' category, and is positive
                                          for all other categories
     "vout" : n,               (numeric) the vout value
     "fee": x.xxx,             (numeric) The amount of the fee in BTC. This is negative and only available for the 'send' category of transactions.
-    "confirmations": n,       (numeric) The number of confirmations for the transaction.
-                                          When it's < 0, it means the transaction conflicted that many blocks ago.
+    "confirmations": n,                        (numeric) The number of confirmations for the transaction. Negative confirmations indicate the
+                                                       transaction conflicts with the block chain.
+    "generated": xxx,                          (bool) Only present if transaction only input is a coinbase one.
+    "trusted": xxx,                            (bool) Only present if we consider transaction to be trusted and so safe to spend from.
     "blockhash": "hashvalue",                  (string) The block hash containing the transaction.
     "blockindex": n,                           (numeric) The index of the transaction in the block that includes it.
     "blocktime": xxx,                          (numeric) The block time in seconds since epoch (1 Jan 1970 GMT).
     "txid": "transactionid",                   (string) The transaction id.
-    "time": xxx,              (numeric) The transaction time in seconds since epoch (Jan 1 1970 GMT).
-    "timereceived": xxx,      (numeric) The time received in seconds since epoch (Jan 1 1970 GMT).
+    "walletconflicts": [                       (array) Conflicting transaction ids.
+      "txid",                                  (string) The transaction id.
+      ...
+    ],
+    "time": xxx,                               (numeric) The transaction time in seconds since epoch (midnight Jan 1 1970 GMT).
+    "timereceived": xxx,                       (numeric) The time received in seconds since epoch (midnight Jan 1 1970 GMT).
+    "comment": "...",                          (string) If a comment is associated with the transaction, only present if not empty.
     "bip125-replaceable": "yes|no|unknown",    (string) Whether this transaction could be replaced due to BIP125 (replace-by-fee);
                                                      may be unknown for unconfirmed transactions not in the mempool
     "abandoned": xxx,         (bool) 'true' if the transaction has been abandoned (inputs are respendable). Only available for the 'send' category of transactions.
     "comment": "...",       (string) If a comment is associated with the transaction.
     "label" : "label"       (string) A comment for the address/transaction, if any
     "to": "...",            (string) If a comment to is associated with the transaction.
   ],
   "removed": [
     <structure is the same as "transactions" above, only present if include_removed=true>
     Note: transactions that were re-added in the active chain will appear as-is in this array, and may thus have a positive confirmation count.
   ],
   "lastblock": "lastblockhash"     (string) The hash of the block (target_confirmations-1) from the best block on the main chain. This is typically used to feed back into listsinceblock the next time you call it. So you would generally use a target_confirmations of say 6, so you will be continually re-notified of transactions until they've reached 6 confirmations plus any new ones
 }
diff --git a/listtransactions b/listtransactions
index 438677e..63249dc 100644
--- a/listtransactions
+++ b/listtransactions
@@ -14,32 +14,38 @@ Arguments:
 Result:
 [
   {
+    "involvesWatchonly": xxx, (bool) Only returns true if imported addresses were involved in transaction.
     "address":"address",    (string) The bitcoin address of the transaction.
     "category":               (string) The transaction category.
                 "send"                  Transactions sent.
                 "receive"               Non-coinbase transactions received.
                 "generate"              Coinbase transactions received with more than 100 confirmations.
                 "immature"              Coinbase transactions received with 100 or fewer confirmations.
                 "orphan"                Orphaned coinbase transactions received.
     "amount": x.xxx,          (numeric) The amount in BTC. This is negative for the 'send' category, and is positive
                                         for all other categories
     "label": "label",       (string) A comment for the address/transaction, if any
     "vout": n,                (numeric) the vout value
     "fee": x.xxx,             (numeric) The amount of the fee in BTC. This is negative and only available for the 
                                          'send' category of transactions.
     "confirmations": n,                        (numeric) The number of confirmations for the transaction. Negative confirmations indicate the
-                                         transaction conflicts with the block chain
-    "trusted": xxx,           (bool) Whether we consider the outputs of this unconfirmed transaction safe to spend.
+                                                       transaction conflicts with the block chain.
+    "generated": xxx,                          (bool) Only present if transaction only input is a coinbase one.
+    "trusted": xxx,                            (bool) Only present if we consider transaction to be trusted and so safe to spend from.
     "blockhash": "hashvalue",                  (string) The block hash containing the transaction.
     "blockindex": n,                           (numeric) The index of the transaction in the block that includes it.
     "blocktime": xxx,                          (numeric) The block time in seconds since epoch (1 Jan 1970 GMT).
     "txid": "transactionid",                   (string) The transaction id.
+    "walletconflicts": [                       (array) Conflicting transaction ids.
+      "txid",                                  (string) The transaction id.
+      ...
+    ],
     "time": xxx,                               (numeric) The transaction time in seconds since epoch (midnight Jan 1 1970 GMT).
     "timereceived": xxx,                       (numeric) The time received in seconds since epoch (midnight Jan 1 1970 GMT).
-    "comment": "...",       (string) If a comment is associated with the transaction.
+    "comment": "...",                          (string) If a comment is associated with the transaction, only present if not empty.
     "bip125-replaceable": "yes|no|unknown",    (string) Whether this transaction could be replaced due to BIP125 (replace-by-fee);
                                                      may be unknown for unconfirmed transactions not in the mempool
     "abandoned": xxx          (bool) 'true' if the transaction has been abandoned (inputs are respendable). Only available for the 
                                          'send' category of transactions.
   }
 ]

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1655,40 +1658,32 @@ static UniValue gettransaction(const JSONRPCRequest& request)
-            "  ],\n"
-            "  \"hex\" : \"data\"         (string) Raw data for transaction\n"
-            "  \"decoded\" : transaction         (json object) Optional, the decoded transaction (only present when `verbose` is passed), equivalent to the\n"
-            "                                                RPC decoderawtransaction method, or the RPC getrawtransaction method when `verbose` is passed.\n"
```
Why is this removed?

ariard replied:
Corrected

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1359,6 +1359,27 @@ static void ListTransactions(interfaces::Chain::Lock& locked_chain, CWallet* con
+           "      ...\n"
+           "    ],\n"
+           "    \"time\": xxx,                               (numeric) The transaction time in seconds since epoch (midnight Jan 1 1970 GMT).\n"
+           "    \"timereceived\": xxx,                       (numeric) The time received in seconds since epoch (midnight Jan 1 1970 GMT).\n"
```
I'd say to remove midnight, because we don't use that nomenclature anywhere else

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1359,6 +1359,27 @@ static void ListTransactions(interfaces::Chain::Lock& locked_chain, CWallet* con
+static const std::string TransactionDescriptionString()
+{
+    return "    \"confirmations\": n,                        (numeric) The number of confirmations for the transaction. Negative confirmations indicate the\n"
+           "                                                       transaction conflicts with the block chain.\n"
```
I liked the previous wording, which was more accurate.

`.. means the transaction conflicted that many blocks ago.`

MarcoFalke commented:
ACK 396da1d8a2ab9c2bd026eece0186b6e9d8c3181c

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 396da1d8a2ab9c2bd026eece0186b6e9d8c3181c
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgcPwv9Gk7KN8Dns6Yj8kdYXui6m+gX2AGVTaROZKRgRXX4ZJao4qrp1cuZABH7
8CRVd8ozqDCPHxupSjxidfDXGvWgiKbhwnfDvNy++wg94UpO9GchUFPAUYg/H/4b
PQIKLN5kE0CgSFI1e4yM6PW/I34VW8Aeh5K1d/PKW7560ZgnHAppOVxDZP3XPNiW
qVYQHDNn/HVGO1G4OORpDr5d7cHbxV92fI1iR0xki2m1hvAlBiwQlDNiBfIvd2hA
mS5PYZHa7NtyvLuLDPiVsY1KWBONXJr9bGfM5mKrdDIOO/Rp+/E0GpladfZGokTy
Y87e3ow1KcrB1hB/wz8NngkC11W1bEievKSQjP34aXJL8DSoO2w3Jj3x0vQxuor3
y8HU6OteEk2uPn0Y0yR2l+6sHdnnrbUKzWdboIgsSvNjW8oZ7lF8gIxf4wQLNG6h
7bc5Itzel/RI8BC2abmLKx9uIUNpLKy/DDDdW8pF3d7qYeGPQX7yxAP835CyGY9z
W4vkc8g5
=/J2v
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `3a1dd9f9738323e22f94da416a4b83f73c9e03313d0758b17e388c2938fda613  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401083a1dd9f9738323e22f94da416a4b83f73c9e03313d0758b17e388c2938fda613f0104b549ecf0b1d8a715013e026cf8615aa08fff01049fc68b6d58f4fd3b490200de0254dce08f120bc3ebd9df981fd6f2d2120856d39301602f6f214f894a5f42aafca5ca8ce549d08f1045d9e1649f0081104623e5b4090b00083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01058771e40b188b91c63b8569bbf45a3a908f120e02264883d600a4a3c0d250130a499af430d2af806000eb11d1f35cb7d6051e808f1045d9e1649f008de27292e92f311dd0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff0107879b47078527c3ee67a0c7c6a2694b108f1045d9e1649f0086364b209646fe00f0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df0108cf86d19e803451736cc48db525b48e408f1045d9e164af008212b3f0000a152360083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

MarcoFalke commented:
re-ACK f4c8953b0073e6bc37abd18ec4a5cbc3ea9719a1 (only change is addressing my nits)

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK f4c8953b0073e6bc37abd18ec4a5cbc3ea9719a1 (only change is addressing my nits)
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhooAwArU21TEvHAToY40eahPE8pCMTN+GSNBaJxtV3/sk2RC6M3h8R+nR78cqa
gN8Qmz4tYENU02ZlnjpfPYjMyh57L8baz/evdb8dzmFd25L59MYBeBggrYWzZ0BF
4mUbl5QEka/tal1dWmoFOK6IR1vVG8CWepH4nyO7qHe3pfeyft+5Dz8IFSG2Uuro
SwhGysluURvuBYZg5aFVazUVSbcuPEzE1EPVRbRJ0CupYwpCnnyabBTdIJrt2nXI
m7vX+301k/wl0zDIyOlK1lRf35C6YTXxgUbmHTrRHrvzVeK42aWUFS2JXJ0MPZW1
ZL4YG3YoLqtfDDVju63E/NOp1rhKIeI8NaHC4qpa8kZeaFoSN1tSdmISgCmftb4P
0rKPyk6Bks5aPNVYTJz9ynuiIuXm599D3o+2D0irx/J/whpk1+EYttAqR0kRjYVv
v5lxfSQ4SeomOnI+hkXkT/Q+ZiMo9TPW6qDQBOo8/1u/hae1Ex9gDqtEA83VBi8J
92CT+k4S
=qvS9
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `fbe29a11ac2d81f0f51edc5354e77b0c147b5e6692a6c02189c7fa41db6287cf  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108fbe29a11ac2d81f0f51edc5354e77b0c147b5e6692a6c02189c7fa41db6287cff0105dee324a16793b5dc6f52fdee47f14bc08fff01050185c6aa43b30e6874e32d5c8aa7f9108f020c3d9e7df133f383b28df3d4a5dc92d696d69d254da34e4c67642a797ed03b07a08f1045d9e3022f0082bb9e78d824fc6b40083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01089d5d4a6466ff7885fbf175ff44705c908f120c733982641f12f8b9a30ecc110fbed7839113cddce633dfaa64a0a68584b5b1508f1045d9e3022f008d9ae28583269832f0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff01093fe2cbe1765c31367299ee1c117a21708f020148beec7d03ac3c39b103a2be50cc0cfeda7dd4974ea7e3c43cb833370cfe0b208f1045d9e3022f008260a6c06e1ea0d0c0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010dfdbe1a2587c9323e3803ee3aa6179f808f1045d9e3022f008156c60f1980462300083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

kristapsk reviewed (approved):
ACK f4c8953b0073e6bc37abd18ec4a5cbc3ea9719a1
